File: android/app/src/main/java/com/k1/gastracker/core/Fx.kt
package com.k1.gastracker.core

import java.time.LocalDate

typealias RatesTable = Map<Triple<LocalDate, String, String>, Double>

sealed class ConvertedCost {
    data class Ready(val amount: Double) : ConvertedCost()
    data object Missing : ConvertedCost()
    data object Unavailable : ConvertedCost()
}

data class FxConversion(
    val costs: Map<Refill, ConvertedCost>,
    val error: String? = null,
    val unavailableCount: Int = 0,
)

fun nearestRateAt(
    rates: RatesTable,
    day: LocalDate,
    fromCurrency: String,
    toCurrency: String,
    maxLookback: Int = 7,
): Double? {
    if (fromCurrency == toCurrency) return 1.0
    rates[Triple(day, fromCurrency, toCurrency)]?.let { return it }
    for (i in 1..maxLookback) {
        rates[Triple(day.minusDays(i.toLong()), fromCurrency, toCurrency)]?.let { return it }
    }
    return null
}

fun convertAmount(amount: Double, rate: Double?): Double? =
    rate?.times(amount)

File: gas_tracker/fx.py
from __future__ import annotations

import datetime as dt


RatesTable = dict[tuple[dt.date, str, str], float]


def nearest_rate_at(
    rates: RatesTable,
    day: dt.date,
    from_currency: str,
    to_currency: str,
    max_lookback: int = 7,
) -> float | None:
    if from_currency == to_currency:
        return 1.0
    key = (day, from_currency, to_currency)
    rate = rates.get(key)
    if rate is not None:
        return rate
    for i in range(1, max_lookback + 1):
        prev = day - dt.timedelta(days=i)
        rate = rates.get((prev, from_currency, to_currency))
        if rate is not None:
            return rate
    return None


def convert_amount(amount: float, rate: float | None) -> float | None:
    if rate is None:
        return None
    return amount * rate
